### DISCUSSION TITLE
Impossible to combine into_boxed and group_by

### DISCUSSION THREAD

**Discussion by dmtrKovalenko** [diesel-rs/diesel#3840] [Q&A]
I have a pretty complicated query with a lot of filters and I need to use the .into_boxed to represent all of them but then I can't use group_by in any form. Here is an example:

 ```rust
 use diesel::PgConnection;
use diesel::QueryDsl;

mod schema {
    diesel::table! {
        users (id) {
            id -> Int4,
            name -> Text,
        }
    }

    diesel::table! {
        comments (id) {
            id -> Int4,
            user -> Int4,
            date -> Date,
            comment -> Nullable<Text>,
            parent -> Nullable<Int4>,
        }
    }

    diesel::joinable!(comments -> users (user));

    diesel::allow_tables_to_appear_in_same_query!(users, comments,);
}

pub fn test_group_by(connection: &mut PgConnection) {
    let query = schema::users::table
        .left_join(schema::comments::table)
        .into_boxed();

    query
        // ERROR: overflow evaluating the requirement `_: diesel::Expression`
        // consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`diesel_demo`)
        .group_by(schema::users::id)
        .load::<(i32, Option<i32>, Option<i32>)>(connection)
        .unwrap();
}
```

It is also impossible to have group by before the into_boxed because

```rust
 let query = schema::users::table
        .left_join(schema::comments::table)
        .group_by(schema::users::id)
        .into_boxed();
```

It fails with a trait validation error: 
```
the trait bound `SelectStatement<FromClause<JoinOn<query_source::joins::Join<users::table, comments::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<comments::columns::user>, NullableExpression<users::columns::id>>>>>, query_builder::select_clause::DefaultSelectClause<FromClause<JoinOn<query_source::joins::Join<users::table, comments::table, LeftOuter>, diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<NullableExpression<comments::columns::user>, NullableExpression<users::columns::id>>>>>>, query_builder::distinct_clause::NoDistinctClause, query_builder::where_clause::NoWhereClause, query_builder::order_clause::NoOrderClause, LimitOffsetClause<NoLimitClause, NoOffsetClause>, query_builder::group_by_clause::GroupByClause<users::columns::id>>: BoxedDsl<'_, _>` is not satisfied
```

**Comment by weiznich** [comment]
I've converted this to a discussion as this did not follow the bug reporting template (and it's obviously not a bug report, but rather a support question).

> I have a pretty complicated query with a lot of filters and I need to use the .into_boxed to represent all of them but then I can't use group_by in any form.

That's "expected" behaviour in so far that it shoudn't compile because it's not supported. It's not supported because we need to know the select clause + the group by clause at compile time for static checking. (What's kind of not expected is the bad errer message there, but that's a compiler problem.)


> It is also impossible to have group by before the into_boxed because

That's also expected behavior as your `GROUP BY` and `SELECT` clause needs to be compatible before boxing. That's not the case here as you mix aggregate (`users::id`) and non-aggregate (all other columns) expressions in your (default) select clause. Set an explicit `.select(_)` clause which is compatible to your `GROUP BY` clause and it will work.

**Reply by dmtrKovalenko** [reply]
Thanks for quick response @weiznich

The really fact is that select statement won't help. The original issue comes from that. I am really surprised it is not included.

**Reply by weiznich** [reply]
The reasons why this is not included are already elaborated in my first answer. If you come up with a better way to check these invariants we are happy to review a feature proposal for that.

**Comment by oulover** [comment]
I had the same problem after upgrading from version 1.4 to version 2.1, but this is supported in version 1.4, is there any alternative?

**Reply by weiznich** [reply]
Just to make one thing clear: It wasn't supported on 1.4.x either, as the officially supported API there did not include `group_by` support. Otherwise the reasoning why this is not supported is already explained [here](https://github.com/diesel-rs/diesel/discussions/3840#discussioncomment-7340994)

**Comment by brunoxylo** [comment]
You can also disable a filter by passing a boolean literal to your query.   
This way you eliminate the need to use `.into_boxed()`   
something like this:   
```rust    
let skip_my_filter :SqlLiteral<Bool> = if my_input.is_empty() {
    sql::<Bool>("TRUE")
} else {
    sql::<Bool>("FALSE")
};
```
... and later in your query you can use that literal to make the filter ineffective   
```rust 
 // if skip_my_filter is true the filter will always evaluate to true 
.filter(my_table::name.eq_any(my_input).or(skip_my_filter))
```
